src/t_hash.c

```diff
@@ -2701,40 +2813,91 @@ void hgetdelCommand(client *c) {
 
     if (server.memory_tracking_per_slot)
         updateSlotAllocSize(c->db, getKeySlot(c->argv[1]->ptr), oldsize, hashTypeAllocSize(o));
-    signalModifiedKey(c, c->db, c->argv[1]);
+    signalModifiedKey(c, c->db, c->argv[keyPos]);
 
     if (expired)
-        notifyKeyspaceEvent(NOTIFY_HASH, "hexpired", c->argv[1], c->db->id);
+        notifyKeyspaceEvent(NOTIFY_HASH, "hexpired", c->argv[keyPos], c->db->id);
+    /* Save key reference before command rewriting */
+    robj *keyObj = c->argv[keyPos];
+
     if (deleted) {
-        notifyKeyspaceEvent(NOTIFY_HASH, "hdel", c->argv[1], c->db->id);
+        notifyKeyspaceEvent(NOTIFY_HASH, "hdel", keyObj, c->db->id);
         server.dirty += deleted;
 
-        /* Propagate as HDEL command.
-         * Orig: HGETDEL <key> FIELDS <numfields> field1 field2 ...
-         * Repl: HDEL <key> field1 field2 ... */
-        rewriteClientCommandArgument(c, 0, shared.hdel);
-        rewriteClientCommandArgument(c, 2, NULL);  /* Delete FIELDS arg */
-        rewriteClientCommandArgument(c, 2, NULL);  /* Delete <numfields> arg */
+        /* Propagate as HDEL command using canonical format */
+        buildCanonicalHdelCommand(c, keyPos, firstFieldPos, num_fields);
     }
 
     /* Key may have become empty because of deleting fields or lazy expire. */
     int64_t newlen = (int64_t) hashTypeLength(o, 0);
     if (newlen == 0) {
         newlen = -1;
         /* Del key but don't update KEYSIZES. else it will decr wrong bin in histogram */
-        dbDeleteSkipKeysizesUpdate(c->db, c->argv[1]);
-        notifyKeyspaceEvent(NOTIFY_GENERIC, "del", c->argv[1], c->db->id);
+        dbDeleteSkipKeysizesUpdate(c->db, keyObj);
+        notifyKeyspaceEvent(NOTIFY_GENERIC, "del", keyObj, c->db->id);
     } else {
         if (hfe && (hashTypeIsFieldsWithExpire(o) == 0)) { /*is it last HFE*/
             estoreRemove(c->db->subexpires, getKeySlot(kvobjGetKey(o)), o);
         }
     }
 
     if (oldlen != newlen)
-        updateKeysizesHist(c->db, getKeySlot(c->argv[1]->ptr), OBJ_HASH,
+        updateKeysizesHist(c->db, getKeySlot(keyObj->ptr), OBJ_HASH,
                            oldlen, newlen);
 }
 
+
+
+/* Helper function to build canonical commands for HGETEX propagation */
+static void buildCanonicalHgetexCommand(client *c, int is_persist, long long expire_time,
```

Comment:
it's only used by hgetex command, i think we can make these code in hgetexCommand directly.